# Querying your data

Pandas provides a whole suite of tools for extracting information from your data. Let's have a look at some of the things that we can do on soe example data:

In [1]:
import pandas as pd

In [2]:
census = pd.read_csv("census.csv")
census

city  year    pop country
0        Paris  2001  2.148  France
1        Paris  2008  2.211  France
2        Paris  2009  2.234  France
3        Paris  2012  2.244  France
4       London  2001  7.322      UK
5       London  2006  7.657      UK
6       London  2011  8.174      UK
7       London  2015  8.615      UK
8         Rome  2001  2.547   Italy
9         Rome  2006  2.627   Italy
10        Rome  2009  2.734   Italy
11        Rome  2012  2.627   Italy
12     Bristol  2000  0.541      UK
13     Bristol  2005  0.578      UK
14     Bristol  2011  0.619      UK
15     Bristol  2015  0.648      UK
16  Birmingham  2001  2.285      UK
17  Birmingham  2006  2.368      UK
18  Birmingham  2010  2.428      UK
19  Birmingham  2015  2.515      UK
20        Lyon  2000  1.445  France
21        Lyon  2005  1.509  France
22        Lyon  2010  1.576  France
23        Lyon  2015  1.646  France

If you just want a peek at the data, you can just grab the first few rows with `head()`. This is useful for getting a quick summary of the columns:

In [3]:
census.head(3)

city  year    pop country
0  Paris  2001  2.148  France
1  Paris  2008  2.211  France
2  Paris  2009  2.234  France

When using `list`s and `dict`s in Python, the square-bracket syntax was used to fetch an item from the container. In Pandas it's a much more powerful tool, as we will see thoughout this chapter.

If you pass a single string to the square brackets of a `DataFrame` it will return to you just that one column:

In [4]:
census["city"]

0          Paris
1          Paris
2          Paris
3          Paris
4         London
5         London
6         London
7         London
8           Rome
9           Rome
10          Rome
11          Rome
12       Bristol
13       Bristol
14       Bristol
15       Bristol
16    Birmingham
17    Birmingham
18    Birmingham
19    Birmingham
20          Lyon
21          Lyon
22          Lyon
23          Lyon
Name: city, dtype: object

Accessing a column like this returns an object called [a `Series`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html) which is the second of the two main Pandas data types. Don't worry too much about these just yet but think of them as being a single column of the `DataFrame`, along with the index of the `DataFrame`.

## Filtering

On of the most common data anlysis taks os to select a subset of your data, based on some condition or property. For example, we might want to get only the census data for Paris and forget about the rest for now.

In Pandas there are two steps to asking a question like this.

1. create a *filter* which describes the quation you want to ask
2. *apply* that filter to the data to get just the bits you are interested in

In [5]:
census["city"] == "Paris"

0      True
1      True
2      True
3      True
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
Name: city, dtype: bool

This has created a new `Series` which has `True` set where the city is Paris and `False` elsewhere.

We can use filtered `Series` like this to filter the `DataFrame` as a whole. `census['city'] == 'Paris'` has returned a `Series` containing booleans. Passing it back into `census` as an indexing operation will use it to filter based on the `'city'` column.

In [6]:
census[census["city"] == "Paris"]

city  year    pop country
0  Paris  2001  2.148  France
1  Paris  2008  2.211  France
2  Paris  2009  2.234  France
3  Paris  2012  2.244  France

You can then carry on and grab another column after that filter:

In [7]:
census[census["city"] == "Paris"]["year"]

0    2001
1    2008
2    2009
3    2012
Name: year, dtype: int64

## Getting rows

If you want to select a **row** from a `DataFrame` then you can use the `.loc` attribute which allows you to pass index values like:

In [8]:
census.loc[2]

city        Paris
year         2009
pop         2.234
country    France
Name: 2, dtype: object

In [9]:
census.loc[2]["city"]

'Paris'

## Adding new columns

New columns can be added to a `DataFrame` simply by assigning them by index (as you would for a Python `dict`) and can be deleted with the `del` keyword in the same way:

In [12]:
census["continental"] = census["country"] != "UK"
census

city  year    pop country  continental
0        Paris  2001  2.148  France         True
1        Paris  2008  2.211  France         True
2        Paris  2009  2.234  France         True
3        Paris  2012  2.244  France         True
4       London  2001  7.322      UK        False
5       London  2006  7.657      UK        False
6       London  2011  8.174      UK        False
7       London  2015  8.615      UK        False
8         Rome  2001  2.547   Italy         True
9         Rome  2006  2.627   Italy         True
10        Rome  2009  2.734   Italy         True
11        Rome  2012  2.627   Italy         True
12     Bristol  2000  0.541      UK        False
13     Bristol  2005  0.578      UK        False
14     Bristol  2011  0.619      UK        False
15     Bristol  2015  0.648      UK        False
16  Birmingham  2001  2.285      UK        False
17  Birmingham  2006  2.368      UK        False
18  Birmingham  2010  2.428      UK        False
19  Birmingham  2015  2.515      UK        False
20        Lyon  2000  1.445  France         True
21        Lyon  2005  1.509  France         True
22        Lyon  2010  1.576  France         True
23        Lyon  2015  1.646  France         True

In [11]:
del census["continental"]

### Exercise

- Create the `DataFrame` containing the census data for the three cities.
- Select the data for the year 2001. Which city had the smallest population that year?
- Find all the cities which had a population smaller than 2.6 million.
- [<small>answer</small>](answer_dataframe_years.ipynb)